designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+| Compile     | ``dotnet lambda package --configuration release --output-package $BUILD/package.zip``  |
++-------------+----------------------------------------------------------------------------------------+
+| Copy Source | No Op                                                                                  |
++-------------+----------------------------------------------------------------------------------------+
+
+
+Python using PIP
+^^^^^^^^^^^^^^^^
+
++-------------+---------------------------------------------------------------------------------------+
+| Action      | Command                                                                               |
++=============+=======================================================================================+
+| Resolve     | ``pip install --isolated --disable-pip-version-check -r requirements.txt -t $BUILD``  |
++-------------+---------------------------------------------------------------------------------------+
+| Compile     | No Op                                                                                 |
++-------------+---------------------------------------------------------------------------------------+
+| Copy Source | Copy all files                                                                        |
++-------------+---------------------------------------------------------------------------------------+
+
+
+Custom Build Commands
+~~~~~~~~~~~~~~~~~~~~~
+
+Users can provide custom build command for each runtime via ``.samrc`` that will be run instead of the default build
+action provided by SAM CLI. SAM CLI will invoke this command for every resource that uses this runtime. This can be any
+arbitrary program that is accessible (via ``PATH`` resolution). The program will be invoked by setting three
+environment variables:
+
+#. ``__SOURCE_DIR`` - Folder where the source files are located
+#. ``__BUILD_DIR`` - Folder where the built artifacts should be written to
+#. ``__RUNTIME`` - AWS Lambda function runtime
+#. ``__TEMPLATE`` - Path to SAM Template
+#. Current Working Directory - Same directory where ``sam build`` command was invoked from
+
+This command is expected to return an exit code of ``zero`` to indicate success. Non-zero exit codes indicate failure.
+
+Building Native Binaries
+~~~~~~~~~~~~~~~~~~~~~~~~
+To build native binaries, we need to run on an architecture and operating system that is similar to AWS Lambda. We use
+the Docker containers provided by `Docker Lambda <https://github.com/lambci/docker-lambda>`_ project to run the same
+set of commands described above on this container. We will mount source code folder and build folder into the container
+so the commands have access to necessary files.
+
+If the user specified a custom build command, we will attempt to run the command as-is within the Docker Container.
+But it is possible that the command might not be available inside the container.
```

Comment:
Called this out above, but how do we ensure that doesn't happen (for at least the default cases)? It's a pretty poor experience if I can build locally w/o a docker container and then can't inside the container. This breaks the experience of build, imo.

Reply:
Agreed. And hence another reason to skip support for running arbitrary build commands in MVP.